enhancements/installer/signal-cluster-deletion.md

```diff
@@ -0,0 +1,130 @@
+---
+title: signal-cluster-deletion
+authors:
+  - "@abutcher"
+reviewers:
+  - dgoodwin
+  - abhinavdahiya
+approvers:
+  - dgoodwin
+  - abhinavdahiya
+creation-date: 2020-03-30
+last-updated: 2020-04-06
+status: provisional
+---
+
+# Signal Cluster Deletion
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Operators can create external resources in clouds and other services that have no opportunity to be cleaned up when a cluster is uninstalled, causing extra work for cluster maintainers to track down and remove these resources. Signalling cluster deletion would allow operators to clean up external resources during uninstallation and would ensure that total cleanup is an automatic process.
+
+## Motivation
+
+### Goals
+
+Signal that cluster deletion has begun through deletion of a new CRD `ClusterAlive` and wait for that object to be deleted during uninstall before continuing with removing cluster resources such as instances, storage and networking.
+
+### Non-Goals
+
+* Attaching finalizers to operator resources based on the `ClusterAlive` object which would facilitate removal of operator resources.
+* Removing resources other than the `ClusterAlive` object during cluster uninstall.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints
+
+#### API
+
+Introduce a new CRD `ClusterAlive` from which operator resources could attach a finalizer initiating cascading delete when the `ClusterAlive` resource is deleted.
+
+```yaml
+apiVersion: v1
+kind: ClusterAlive
+metadata:
+  name: cluster
+spec:
+  blockTeardown: true  # block teardown of operator resources
+```
+
+#### Install
+
+Create a `ClusterAlive` object during installation. (Is a manifest the right place to create for new clusters ?)
+
+#### Upgrade
+
+Create `ClusterAlive` object for existing clusters. (Is this the right place to create for existing clusters ?)
+
+#### Uninstall
+
+Delete a cluster's `ClusterAlive` resource during uninstall when requested via flag such as `openshift-install destroy cluster --cluster-alive-delete`. Cluster destroy will wait for a default amount of time and fail if `ClusterAlive` deletion was not successful. The default timeout will not be configurable and users will be expected to attempt shutdown multiple times upon failure.
```

Comment:
The flag will probably not be `--cluster-alive-delete` but I'm worried about making it too generic upfront so as not to suggest the cluster is being shut down, etc.

Reply:
Why would this live behind a flag, instead of being attempted on every `destroy cluster`?  If a cluster is too old to have the custom resource, or a previous attempt already removed the cluster resource, we'd start in on the current installer teardown logic, which seems appropriate.  If you had a flag, under what circumstances would we recommend a user not set the flag?

Reply:
> currently users expect the destroy to work, even in cases where the cluster api is not running. with your proposal the usres will now error if the api is unavailable, or this new delete fails. to me that sounds like backward-incompatible change.

Placing behind a flag was proposed as a way to make this backwards compatible to combat situations where the API is down and we expect the uninstall to succeed no matter what. Knowing when to use the flag would come down to knowing if your operators are creating resources that could be torn down by the two stage delete which may not be apparent to admins.